server/server/views/event.py

```diff
@@ -91,19 +111,97 @@ def collection_post(self):
+            response['status'] = model_to_dict(self.event_history)
+
+        return response
+
+    @view(schema=EventSchema(), validators=(colander_body_validator,),
+          permission='update')
+    def put(self):
+        """Get validated data and update event data"""
+        request = self.request
+        response = {
+            'success': False,
+            'errors': [{
+                'name': 'name',
+                'description': 'Event name already exist'
+            }],
+        }
+        current_event = Event.get_event_by_id(request, self.event_id)
+
+        if current_event is None:
+            raise HTTPNotFound()
+
+        if (current_event.name == request.validated['name'] or
+            Event.get_event_obj(request,
+                                name=request.validated['name']) is None):
+            event_dict = model_to_dict(current_event)
+            for key in event_dict:
+                if key in request.validated:
+                    event_dict[key] = request.validated[key]
+
+            category_obj = Category\
+                .get_by_name(request, request.validated['category'].lower())
+            if category_obj is None:
+                response['errors']['name'] = 'category'
+                response['errors']['description'] = "Category is not exist"
+                return response
+            event_dict['category_id'] = category_obj.id
+
+            current_event.update_event(request, event_dict)
+
+            currnet_event_tags = EventTag.get_event_tags(request,
+                                                         self.event_id)
+            for tag in currnet_event_tags:
+                if tag not in request.validated.get('tags', []):
+                    EventTag.delete_tag(request,
+                                        currnet_event_tags[tag], self.event_id)
+            for tag in request.validated.get('tags', []):
+                tag_obj = Tag.get_by_name(request, tag.lower())
+                if tag_obj is not None:
+                    tag_id = tag_obj.id
+                    if tag not in currnet_event_tags:
+                        EventTag.add_new(request, tag_id, self.event_id)
+                else:
+                    tag_id = Tag.add_new(request, tag.lower())
+                    EventTag.add_new(request, tag_id, self.event_id)
+
+            response['success'] = EventHistory.create_new(
+                request,
+                event_id=self.event_id,
+                status_id=EventStatus.get_status(request, 'New').id,
+                date=datetime.now(),
+                comment=("Your event updated. "
+                         "Please wait for review by moderator.")
+            )
+            del response['errors']
+        return response
+
+    @view(permission='update')
+    def delete(self):
+        """Make event closed"""
+        response = {'success': False}
+        response['success'] = EventHistory.create_new(
```

Comment:
The same as above,

Reply:
The main idea here to hide logic in Model, and controller may even doesn't know anything about EventHistory, as here we deal only with Events.